### PR TITLE
Create `savanna_cluster` for unittests, and use it to port a forked_test.

### PR DESCRIPTION
**Pull request by greg7mdp**
This PR: 

- adds a new class `savanna_cluster::cluster_t` in `unittests` which implements a cluster of four nodes (four `tester` instances) which, by default, automatically propagate blocks and votes to each other. For the purpose of creating forks, the cluster can be split using the `set_partition()` method, and if the partition is comprised of two nodes, finality will stop advancing (as the quorum is 3 finalizers, and we have one finalizer per node).
- provides a test for this new cluster (in `savanna_cluster.cpp`)
- ports the test ` fork_with_bad_block` from `forked_tests.cpp` to Savanna, in the new file `forked_tests_if.cpp`, which will be completed with further ported and new tests in https://github.com/AntelopeIO/spring/issues/10.
- adds a flag allowing votes to be processed synchronously in controller.cpp (`disable_async_voting`) and updates the `tester` to initialize this flag to `true`. This removes the need for the `sleep()` in `base_tester::_check_for_vote_if_needed` method, as when this flag is set and  `controller::push_block()` or `controller::commit_block()` return, we know that the vote(s) (if there was to be one or more) were created and the `voted_block` signal emited.
- split the signal `voted_block` into two separate signals:
    1. `voted_block` which is emited when a local finalizer votes on a block.
    2. `aggregated_vote` which is emited when a vote from a remote finalized was receid and aggregated into a QC.
- a boolean was added to the tester (`_expect_votes`) which will by default be set to true and will trigger an assert if the node has an active finalizer and we don't get a vote when pushing or producing a block. 
There will be a `base_tester::do_check_for_votes()` method allowing to turn this flag off, which will be necessary when creating tests where forks are created.
- add a `base_tester::get_votes(const block_id_type& id)` to retrieve the current votes in the `pending_qc` for a block. This will allow to verify that we get the votes we expect in various tests.

### CODE REVIEW

File: libraries/testing/tester.cpp [hunk 1]
```diff
@@ -508,18 +520,20 @@ namespace eosio::testing {
...
+      _wait_for_vote_if_needed(*control, sb);
 
-      return control->head_block();
+      return sb;
    }
 
-   void base_tester::_wait_for_vote_if_needed(controller& c) {
-      if (c.can_vote_on(c.head_block())) {
+   void base_tester::_wait_for_vote_if_needed(controller& c, const signed_block_ptr& b) {
+      if (0 && c.can_vote_on(b)) {
```
Comment (heifner) [inline libraries/testing/tester.cpp:532]: Remove this temporary `0 &&`

Comment (greg7mdp) [reply libraries/testing/tester.cpp]: done.

**Comment by ericpassmore** [comment]
Note:start
group: STABILITY
category: TEST
summary: Create new cluster for Savanna Unit Test.
Note:end

File: libraries/chain/block_state.cpp [hunk 2]
```diff
@@ -215,6 +223,7 @@ void block_state::verify_qc(const valid_quorum_certificate& qc) const {
    // utility to accumulate voted weights
    auto weights = [&] ( const vote_bitset& votes_bitset ) -> uint64_t {
       uint64_t sum = 0;
+      assert(num_finalizers == votes_bitset.size());
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/block_state.cpp:226]: To demonstrate that the `std::min` on the next line is not needed.

File: libraries/chain/controller.cpp [hunk 3]
```diff
@@ -958,6 +958,8 @@ struct controller_impl {
    deep_mind_handler*              deep_mind_logger = nullptr;
    bool                            okay_to_print_integrity_hash_on_stop = false;
    bool                            allow_voting = true; // used in unit tests to create long forks or simulate not getting votes
+   bool                            disable_async_voting = false; // by default we post `create_and_send_vote_msg()` calls
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/controller.cpp:961]: I added `used in tester`. 

> Any way to ensure it is really disabled in production nodeos?

Not at this point. We don't have that for other tester things like `allow_voting` I believe.

File: plugins/net_plugin/net_plugin.cpp [hunk 4]
```diff
@@ -537,7 +537,7 @@ namespace eosio {
 
       void on_accepted_block_header( const signed_block_ptr& block, const block_id_type& id );
       void on_accepted_block( const signed_block_ptr& block, const block_id_type& id );
-      void on_voted_block( uint32_t connection_id, vote_status stauts, const vote_message_ptr& vote );
+      void broadcast_vote_message( uint32_t connection_id, vote_status stauts, const vote_message_ptr& vote );
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply plugins/net_plugin/net_plugin.cpp:540]: Because the same function is used for two different signals, a `on_` name probably isn't the best choice. I think it makes more sense here to use a name describing what the function does, rather than what signals it responds to.

File: unittests/forked_tests.cpp [hunk 5]
```diff
@@ -118,14 +124,15 @@ BOOST_AUTO_TEST_CASE( fork_with_bad_block ) try {
    }
 
    // make sure we can still produce a blocks until irreversibility moves
-   auto lib = bios.control->head_block_state_legacy()->dpos_irreversible_blocknum;
+   auto lib = bios.lib_block->block_num();
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/forked_tests.cpp:127]: Because the lib is now immediately available in the tester (updated from the ` irreversible_block` signal) I thought it made more sense to use it rather than quering the controller.

File: unittests/forked_tests_if.cpp [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+   struct fork_tracker {
+      vector<signed_block_ptr>           blocks;
+   };
+
+   node0.produce_block();
+
+   const vector<account_name> producers {"a"_n, "b"_n, "c"_n, "d"_n, "e"_n};
+   auto prod = set_producers(0, producers);   // set new producers and produce blocks until the switch is pending
+
+   auto sb = node0.produce_block();           // now the next block produced on any node
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/forked_tests_if.cpp:33]: Right, I wasn't clear, I meant it can be produced on any node. I updated the comment.

File: unittests/savanna_cluster.hpp [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+         : _nodes{
+               {{0, *this, setup_policy::full_except_do_not_transition_to_savanna}, {1, *this}, {2, *this}, {3, *this}}
+      } {
+         // make sure we push node0 initialization (full_except_do_not_transition_to_savanna) to
+         // the other nodes. Needed because the tester was initialized before `node_t`.
+         // ------------------------------------------------------------------------------------
+         for (size_t i = 0; i < _nodes.size(); ++i)
+            node0.push_blocks(_nodes[i]);
+
+         // from now on, propagation of blocks and votes happens automatically.
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp:116]: Thanks, I updated the comment and added that info.

File: unittests/forked_tests_if.cpp [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+//    - irreversible_mode
+//    - push_block_returns_forked_transactions
+//
+// Similar Legacy tests are in: `forked_tests.cpp`
+// ---------------------------------------------------
+
+BOOST_AUTO_TEST_SUITE(forked_tests_if)
+
+// ---------------------------- fork_with_bad_block ---------------------------------
+BOOST_FIXTURE_TEST_CASE(fork_with_bad_block_if, savanna_cluster::cluster_t) try {
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/forked_tests_if.cpp:36]: Sure, done.

File: unittests/savanna_cluster.hpp [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+         while (1) {
+            sb = n.produce_block();
+            pending = n.control->pending_block_producer();
+            if (ranges::any_of(producers, [&](auto a) { return a == pending; }))
+               break;
+         }
+         return ranges::find(producers, pending) - producers.begin();
+      }
+
+      // provide a set of node indices wgich will be disconnected from other nodes of the network,
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp]: updated.

File: libraries/testing/tester.cpp [hunk 10]
```diff
@@ -353,6 +353,9 @@ namespace eosio::testing {
           }
       });
 
+      control->disable_async_voting(true);      // vote synchronously so we don't have to wait for votes
```
Comment (linh2931) [inline libraries/testing/tester.cpp:356]: Should this be set to `false` by default? Only let specific tests (like fork tests) set it to `true` when needed?

Comment (greg7mdp) [reply libraries/testing/tester.cpp:356]: I think it is better to have async voting disabled by default. I don't see why any test would need asynchronous voting.

File: libraries/chain/include/eosio/chain/vote_processor.hpp [hunk 11]
```diff
@@ -185,27 +185,33 @@ class vote_processor_t {
    }
 
    // called from net threads
-   void notify_new_block() {
+   void notify_new_block(bool disable_async = false) {
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/include/eosio/chain/vote_processor.hpp:188]: @heifner since you originally wrote this, I wonder what you prefer?

[the remark this answers is not in the document]
Comment (heifner) [reply libraries/chain/include/eosio/chain/vote_processor.hpp:188]: Prefer an enum, non-defaulted

[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/include/eosio/chain/vote_processor.hpp:188]: I assume something like:

```
enum class async_t { off, on };
 void notify_new_block(async_t async);
```

[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/include/eosio/chain/vote_processor.hpp]: done

File: libraries/testing/tester.cpp (hunk 10 above)
Comment (linh2931) [reply libraries/testing/tester.cpp]: Asynchronous voting is used in production. It is better to leave some tests exercising it.

**Review by linh2931 — APPROVED** [review]

File: unittests/forked_tests.cpp [hunk 12]
```diff
@@ -3,22 +3,27 @@
...
+//    - push_block_returns_forked_transactions
+//
+// Similar Savanna tests are in: `forked_tests_if.cpp`
+// ---------------------------------------------------
+
 BOOST_AUTO_TEST_SUITE(forked_tests)
 
+// ---------------------------- irrblock ---------------------------------
 BOOST_AUTO_TEST_CASE( irrblock ) try {
    legacy_tester c;
```
Comment (arhag) [inline unittests/forked_tests.cpp:28]: This is also a test that appears to only use `legacy_tester` but is not in the list in your comment above. But, this `irrblock` test case (which is also badly named) is so pointless that I am fine with just deleting it.

File: unittests/forked_tests_if.cpp [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+            // push the block only if its not known already
+            if (!node0.control->fetch_block_by_id(b->calculate_id())) {
+               node0.push_block(b);
+            }
+         }
+
+         // push the block which should attempt the corrupted fork and fail
+         BOOST_REQUIRE_EXCEPTION( node0.push_block(fork.blocks.back()), fc::exception,
+                                  fc_exception_message_starts_with( "finality_mroot does not match" )
+         );
```
Comment (arhag) [inline unittests/forked_tests_if.cpp:142]: Also, after each failure, check to see that the head block of node0 is the same as what it was before the test started pushing any of the bad forks.

File: unittests/forked_tests_if.cpp [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+   }
+
+   // make sure we can still produce blocks until irreversibility moves
+   // -----------------------------------------------------------------
+   auto lib = node0.lib_block->block_num();
+   size_t tries = 0;
+   while (node0.lib_block->block_num() == lib && ++tries < 10) {
+      node0.produce_block();
+   }
+
```
Comment (arhag) [inline unittests/forked_tests_if.cpp:153]: Check that the lib has in fact advanced and we didn't just exit the while loop because `tries` reached the limit.

**Review by arhag — CHANGES_REQUESTED** [review]

File: libraries/chain/block_state.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/block_state.cpp:226]: > Do we reject the QC pulled from a block if the size is not the same as the number of finalizers for the policy it refers to?
> 
> I want to make sure a bad block wouldn't trip this assertion and would instead trip an earlier check that would throw and exception an reject the block.

There is no earlier check for this that would reject the block. This finction here (`block_state::verify_qc`) is the only check we do on the QC itself.

I added an assert which is enabled only on debug builds, so it doesn't change spring's behavior. Currently because of the `std::min` on line 227 we will accept the QC, but may not consider all votes if there is a mismatch in size. It did seem odd to me that we would quiely accept a size mismatch.

According to what you write, it sounds like  we should change the `assert` into a `EOS_ASSERT` so the block is rejected.

File: libraries/testing/tester.cpp [hunk 15]
```diff
@@ -508,21 +516,23 @@ namespace eosio::testing {
       } );
 
       control->commit_block(br);
-      last_produced_block[producer_name] = control->head_block_id();
+      signed_block_ptr sb = control->head_block();
+      last_produced_block[producer_name] = sb->calculate_id();
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/testing/tester.cpp:520]: Sure will revert to `control->head_block_id()`.

File: unittests/savanna_cluster.hpp [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+      // returns the number of nodes on which `lib` advanced since we last checked
+      // -------------------------------------------------------------------------
+      size_t num_lib_advancing() {
+         return ranges::count_if(_nodes, [](node_t& n) { return n.lib_advancing(); });
+      }
+
+      void reset_lib() { for (auto& n : _nodes) n.reset_lib();  }
+
+   public:
+      std::array<node_t, num_nodes>   _nodes;
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp:191]: I was thinking of renaming node0, ... to `A`, `B`, `C` `D` as used in the cryptolions document.

File: unittests/savanna_cluster.hpp [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+   //
+   // By default they are all connected, receive all produced blocks, vote on them,
+   // and send their votes to all other nodes.
+   //
+   // It is possible to split the 'virtual' network using `cluster_t::split()`.
+   //  --------------------------------------------------------------------------------------
+   class cluster_t {
+   public:
+      static constexpr size_t num_nodes = 4;
+      static constexpr size_t keys_per_node = 10;
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp:98]: Well, I did make the `finality_test_cluster` a template class which takes `NUM_NODES` as a template parameter. I felt that it was less needed for the `savanna_cluster`, but yes probably  I didn't envision all the possible use cases so I'll change it. 

Do you think there should be a minimal number of nodes? With one node you don't create branches and finality always advances. With two or 3 nodes you need all the nodes to vote for finality to advance. 4 seemed like a nice number where the quorum is 3. 

As of now, `finality_test_cluster` requires a minimum of 4 nodes.

[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp:98]: > It would be great if the number of nodes were dynamic and determined by an argument to the cluster. 

Would you mind if I made the cluster a `template<size_t NUM_NODES> ` instead of the cluster taking an argument? It would allow to keep the array of nodes as a `std::array`.

File: unittests/savanna_cluster.hpp [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+      node_t&                         node0 = _nodes[0];
+      node_t&                         node1 = _nodes[1];
+      node_t&                         node2 = _nodes[2];
+      node_t&                         node3 = _nodes[3];
+
+      // Used for transition to Savanna
+      // ------------------------------
+      std::optional<finalizer_policy> _fin_policy_0;         // policy used to transition to Savanna
+      std::array<size_t, num_nodes>   _fin_policy_indices_0; // set of key indices used for transition
+      std::vector<bls_public_key>     _fin_policy_pubkeys_0; // set of public keys used for transition
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp:202]: It is not important, I'll remove these.

File: unittests/savanna_cluster.hpp [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+      }
+
+      void push_block_to_peers(size_t node_idx, const signed_block_ptr& b) {
+         for_each_peer(node_idx, [&](node_t& n) {
+            n.push_block(b);
+         });
+      }
+
+      template<class CB>
+      void for_each_peer(size_t node_idx, const CB& cb) {
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp:224]: I feel that the benefit (most likely unmeasurable compilation speed improvement, even if `savanna_cluster.hpp` was included in a couple dozen test.cpp files) does not make up for the added complexity/decreaded readibility, but I can certainly make the change.

To support the claim that it wouldn't make a difference in compilation speed, I checked that the preprocessed size of `forked_tests_if.cpp` is 412,028 lines (thanks to the boost headers), so the 170 lines of `savanna_cluster.cpp` barely register.

File: unittests/forked_tests.cpp (hunk 12 above)
Comment (greg7mdp) [reply unittests/forked_tests.cpp:28]: Sounds good, removing it.

File: unittests/forked_tests.cpp [hunk 20]
```diff
@@ -333,19 +309,21 @@ BOOST_AUTO_TEST_CASE( prune_remove_branch ) try {
       c.push_block(fb);
    }
 
-   BOOST_REQUIRE_EQUAL(73u, c.control->head_block_num());
+   BOOST_REQUIRE_EQUAL(fork_num + 24u, c.control->head_block_num());
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/forked_tests.cpp:312]: Oh I completely missed that I had changed the behavior of the test! Thanks for noticing that.

[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/forked_tests.cpp]: I fixed what we produce before `push_blocks`, so that we match the old version of the test. Now it does pick (and checks) the shorter fork. Thanks again for pointing my mistake. I hope my comment after `c.produce_block(); ` is accurate.

File: unittests/forked_tests.cpp [hunk 21]
```diff
@@ -376,12 +354,17 @@ BOOST_AUTO_TEST_CASE( validator_accepts_valid_blocks ) try {
...
 
    BOOST_CHECK_EQUAL( n3.control->head_block_id(), id );
-
-
 } FC_LOG_AND_RETHROW()
 
-BOOST_AUTO_TEST_CASE( read_modes ) try {
-   legacy_tester c;
+BOOST_AUTO_TEST_CASE( validator_accepts_valid_blocks ) {
+   legacy_tester c;		   test_validator_accepts_valid_blocks<legacy_tester>();
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/forked_tests.cpp:352]: done

File: unittests/forked_tests_if.cpp (hunk 13 above)
Comment (greg7mdp) [reply unittests/forked_tests_if.cpp:142]: done.

File: unittests/forked_tests_if.cpp (hunk 14 above)
Comment (greg7mdp) [reply unittests/forked_tests_if.cpp:153]: OK, I have a problem here. For finality to advance, I need 3 finalizers voting, but nodes 2 and 3 have bad heads, so when node0 produces a block and attempts to push it to node2 or node3 I have a `unlinkable_block_exception`.

How does a node recover from having a bad head fork?

File: unittests/savanna_cluster.hpp [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+         _shutting_down = true;
+      }
+
+      // Create accounts and updates producers on node node_idx (producer updates will be
+      // propagated to connected nodes), and wait until one of the new producers is pending.
+      // return the index of the pending new producer (we assume no duplicates in producer list)
+      // -----------------------------------------------------------------------------------
+      size_t set_producers(size_t node_idx, const std::vector<account_name>& producers) {
+         node_t& n = _nodes[node_idx];
+         n.create_accounts(producers);
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp]: done

File: libraries/testing/tester.cpp [hunk 23]
```diff
@@ -508,21 +516,23 @@ namespace eosio::testing {
...
-         FC_ASSERT(retrys, "Never saw this nodes vote processed before timeout");
+   void base_tester::_check_for_vote_if_needed(controller& c, const signed_block_ptr& b) {
+      if (_expect_votes && c.can_vote_on(b)) {
+         // `_expect_votes` should be true *only* when we expect an active finalizer to
+         // vote on every block.
+         // This is not the case for tests with forks, so for these tests we should set
+         // `_expect_votes` to false by calling `base_tester::do_check_for_votes(false)`
+         // ----------------------------------------------------------------------------
+         auto block_id = b->calculate_id();
+         FC_ASSERT(c.node_has_voted_if_finalizer(block_id), "Missing expected vote");
```
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/testing/tester.cpp]: done.

> The only tricky part would be getting getting the `block_handle` in `_finish_block`.

I added a `block_handle controller::head() const { return my->chain_head; }` member in `controller`.

Now that we have a `head()` member returning a `block_handle`, I would be in favor of removing the other accessors below it in the following link: https://github.com/AntelopeIO/spring/blob/df1c5ac8d7059ef05c1bd7feaff9bfefcc4862eb/libraries/chain/controller.cpp#L5095-L5118

File: unittests/savanna_cluster.hpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp]: OK, made `savanna_cluster` a template:

```
template <size_t NUM_NODES, size_t KEYS_PER_NODE = 10> requires (NUM_NODES > 3)
class cluster_t {
...
```

File: unittests/forked_tests_if.cpp (hunk 14 above)
Comment (greg7mdp) [reply unittests/forked_tests_if.cpp]: OK, this is addressed, I added a `propagate_heads` member to `savanna_cluster`.

File: libraries/testing/tester.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (arhag) [reply libraries/testing/tester.cpp]: Looks great!

I think I agree with removing the other accessors which are made redundant by simply providing `head()`.
My only hesitation is that `head()` exposes the `block_handle` to components other than the internals of libchain and tester, e.g. plugins, and that wouldn't necessarily be a problem as long as it does not expose the internal `block_header_state`.

But `block_handle::internal()` return the internal `_bsp`. Maybe the name is sufficient to indicate to others that this is not part of the API that they should rely on. But maybe there is a way to use `protected` and `friend` to enforce better restrictions so that the compiler only allows the the code in `controller.cpp` to access the internal block header state via the `block_handle`?

File: unittests/forked_tests.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (arhag) [reply unittests/forked_tests.cpp]: Looks good!

File: libraries/chain/block_state.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (arhag) [reply libraries/chain/block_state.cpp:228]: Yes, it appears instead of this `assert` we need `EOS_ASSERT`s in `verify_qc` to verify that both the `strong_votes` and `weak_votes` if present are of the right size (they match the number of finalizers) otherwise we reject the QC as invalid. And there should be test cases to demonstrate that check works.

File: unittests/savanna_cluster.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (arhag) [reply unittests/savanna_cluster.hpp]: I don't understand why `fin_policy_pubkeys` is needed either?

Don't we already have a way to verify if the relevant finalizers of each node have voted on a particular block using the `is_block_missing_finalizer_votes` function?

File: unittests/savanna_cluster.hpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (arhag) [reply unittests/savanna_cluster.hpp]: No, I think we should avoid the letter aliases, or any aliases whatsoever. `node[n]` is pretty straightforward already.

File: unittests/savanna_cluster.hpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (arhag) [reply unittests/savanna_cluster.hpp]: I don't see a strong reason to advocate for using a `std::array` over a `std::vector` in this case. It is just one allocation during the constructor, and performance of an extra indirection doesn't really matter in this case.

And on the other hand, I think it makes the code a little more complicated by making it a template. In particular, making the the node class into a template so that it can take the now template cluster class.

By the way the constructor of the cluster is still assuming exactly four nodes despite making `NUM_NODES` a template paramater. I think it is fine to enforce a minimum value of 4 for `num_nodes`. But I do think this would all be a little cleaner if it was just provided as a constructor argument in a class that is not a template.

Also, I do not see why `keys_per_node` is required at all? In fact, I find the `finalizer_keys` template class to be hard to follow in general. So far, I am not seeing why you need more than one finalizer key per node. I can see how there could be value for that for some tests where you want to rotate the finalizer key for a node. But even then I don't see why you would need to initial a fixed number of finalizer per node rather than just generation finalizer keys as needed.

For example, https://github.com/AntelopeIO/spring/blob/main/libraries/testing/include/eosio/testing/bls_utils.hpp could be modified to take a `sha256` hash directly as the seed to generate the private key, public key, and proof of possession rather than arbitrarily limiting oneself only to a `eosio::chain::name`. Then the tester utility could generate a unique key (within the cluster) by tracking a finalizer key sequence number per each node and generating a `sha256` hash to use as the seed from a SHA256 hash of both the node index and the finalizer key sequence number.

File: unittests/savanna_cluster.hpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (arhag) [reply unittests/savanna_cluster.hpp:271]: Yes, I don't think performance was a strong justification in this case for template vs no template or for implementation in separate file from interface vs unified in one header file. I was pushing for it more in terms of readability and simplicity. I think avoiding the template nature of the overall class, where it isn't really necessary simplifies the code a little bit. But even if going with a no template approach (other than for the callback functions), you could still keep the implementation all in one place. Personally, I think it makes it a little more noisy when just trying to scan through the interface when deciding how to use the classes (and not caring about how exactly they are implemented). But I also do not have much of a strong preference either way in this case.

File: libraries/testing/tester.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/testing/tester.cpp]: > But maybe there is a way to use protected and friend to enforce better restrictions so that the compiler only allows the the code in controller.cpp to access the internal block header state via the block_handle?

Just implemented this in latest commit https://github.com/AntelopeIO/spring/pull/316/commits/978e88864dcab02eecf7de504f2f2e9fdace3465

File: unittests/savanna_cluster.hpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp]: OK, if that's OK I'll change that in my other branch `gh_10` (for issue #10) as that other branch is based on this one and changing here would be more of a hassle.

File: unittests/savanna_cluster.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp]: Yes, it is true, I could use 

Right now I use `base_tester::get_votes()`: 

https://github.com/AntelopeIO/spring/blob/978e88864dcab02eecf7de504f2f2e9fdace3465/unittests/savanna_cluster_tests.cpp#L30-L35

which has the advantage of taking a block id and does a lookup in fork_db.

`is_block_missing_finalizer_votes` takes a `block_handle` as returned by `head()`, but I could add a version that takes a block_id and does the lookup in fork_db as well.

File: unittests/savanna_cluster.hpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp]: A problem is that when using `BOOST_FIXTURE_TEST_CASE `, you provide a parent class, but as far as I can tell you cannot pass an argument to the class constructor. So I feel that if we want to support a variable number of nodes, the template is actually a better solution here.

[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp]: > By the way the constructor of the cluster is still assuming exactly four nodes despite making `NUM_NODES` a template paramater.

You are right, I may have to switch to a vector anyways to initialize the nodes correctly.
I'll do that in the other branch `gh_10`, because if I switch to a vector, I cannot have the references `node0`, `node1` declared as member in the class, and I'll remove these in `gh_10`.

File: libraries/chain/block_state.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (greg7mdp) [reply libraries/chain/block_state.cpp:228]: Created issue #334 (because of the test cases)

File: unittests/savanna_cluster.hpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (arhag) [reply unittests/savanna_cluster.hpp]: > A problem is that when using BOOST_FIXTURE_TEST_CASE , you provide a parent class, but as far as I can tell you cannot pass an argument to the class constructor. So I feel that if we want to support a variable number of nodes, the template is actually a better solution here.

A typical way of handling this is to create a wrapper derived class which uses its default constructor to call the base constructor with different values.

For convenience, the `savanna_cluster` could have default values for the parameters of its constructor so that it can be used directly in most tests that do not need to change the number of nodes.

> You are right, I may have to switch to a vector anyways to initialize the nodes correctly.
I'll do that in the other branch gh_10, because if I switch to a vector, I cannot have the references node0, node1 declared as member in the class, and I'll remove these in gh_10.

Yes, I think the `node0`, `node1`, etc. should be removed. I would rather see the incorrect initialization fixed in this PR since as the code currently exists it does not actually work correctly with `NUM_NODES` other than 4.

[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp]: OK, doing it. There is the same pattern (`template<size_t NUM_NODES>`, `node0` and `node1` aliases) in `finality_test_cluster.hpp`. Should I update this one as well for consistency?

[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp]: I updated `savanna_cluster` to not be a template anymore, use a vector for nodes and forego the aliases.

File: unittests/savanna_cluster.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (greg7mdp) [reply unittests/savanna_cluster.hpp]: I updated to not store the finalizer keys anymore, and use `is_block_missing_finalizer_votes` instead.

**Review by arhag — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: STABILITY
category: TEST
summary: New and Improved testing infrastructure, Savanna Cluster to support unit tests and fork tests.
Note:end
